internal/pkg/peer/blocksprovider/deliverer.go

```diff
@@ -142,196 +165,74 @@ func (d *Deliverer) DeliverBlocks() {
 			return
 		}
 
-		seekInfoEnv, err := d.createSeekInfo(ledgerHeight)
+		endpoint, err := d.Orderers.RandomEndpoint()
+		if err != nil {
+			d.Logger.Warningf("Could not connect to ordering service: could not get orderer endpoints: %s", err)
+			failureCounter++
+			continue
+		}
+
+		seekInfoEnv, err := d.requester.SeekInfoBlocksFrom(ledgerHeight)
 		if err != nil {
 			d.Logger.Error("Could not create a signed Deliver SeekInfo message, something is critically wrong", err)
 			return
 		}
 
-		deliverClient, endpoint, cancel, err := d.connect(seekInfoEnv)
+		deliverClient, cancel, err := d.requester.Connect(seekInfoEnv, endpoint)
 		if err != nil {
 			d.Logger.Warningf("Could not connect to ordering service: %s", err)
 			failureCounter++
 			continue
 		}
 
-		connLogger := d.Logger.With("orderer-address", endpoint.Address)
-		connLogger.Infow("Pulling next blocks from ordering service", "nextBlock", ledgerHeight)
-
-		recv := make(chan *orderer.DeliverResponse)
-		go func() {
-			for {
-				resp, err := deliverClient.Recv()
-				if err != nil {
-					connLogger.Warningf("Encountered an error reading from deliver stream: %s", err)
-					close(recv)
-					return
-				}
-				select {
-				case recv <- resp:
-				case <-d.DoneC:
-					close(recv)
-					return
-				}
-			}
-		}()
-
-	RecvLoop: // Loop until the endpoint is refreshed, or there is an error on the connection
-		for {
-			select {
-			case <-endpoint.Refreshed:
-				connLogger.Infof("Ordering endpoints have been refreshed, disconnecting from deliver to reconnect using updated endpoints")
-				break RecvLoop
-			case response, ok := <-recv:
-				if !ok {
-					connLogger.Warningf("Orderer hung up without sending status")
-					failureCounter++
-					break RecvLoop
-				}
-				err = d.processMsg(response)
-				if err != nil {
-					connLogger.Warningf("Got error while attempting to receive blocks: %v", err)
-					failureCounter++
-					break RecvLoop
-				}
-				failureCounter = 0
-			case <-d.DoneC:
-				break RecvLoop
-			}
+		d.mutex.Lock()
+		blockReceiver := &BlockReceiver{
+			channelID:      d.ChannelID,
+			blockHandler:   d.BlockHandler,
+			blockVerifier:  d.BlockVerifier,
+			deliverClient:  deliverClient,
+			cancelSendFunc: cancel,
+			recvC:          make(chan *orderer.DeliverResponse),
```

Comment:
this is a non buffered channel. so we ingest a block one by one. 

In gossip there is a buffer, but in the orderer there isn't one, so this is not efficient to pull blocks one by one and commit them one by one.

Reply:
The plan is to have a buffer at the orderer as well, a buffer that gets all the blocks taken from this channel. The follower/chain goroutines then pull blocks from that buffer, not directly from the `recvC` channel shown here.